packages/compiler/emitter-framework/asset-emitter.ts

```diff
@@ -0,0 +1,749 @@
+    context = newContext.context;
+    lexicalTypeStack = newContext.lexicalTypeStack;
+
+    cb();
+
+    context = oldContext;
+    lexicalTypeStack = oldTypeStack;
+  }
+
+  function typeEmitterKey(type: Type) {
+    switch (type.kind) {
+      case "Model":
+        if (type.name === "" || type.name === "Array") {
+          return "modelLiteral";
+        }
+
+        if (type.templateMapper) {
+          return "modelInstantiation";
+        }
+
+        return "modelDeclaration";
+      case "Namespace":
+        return "namespace";
+      case "ModelProperty":
+        return "modelPropertyLiteral";
+      case "Boolean":
+        return "booleanLiteral";
+      case "String":
+        return "stringLiteral";
+      case "Number":
+        return "numericLiteral";
+      case "Operation":
+        if (type.interface) {
+          return "interfaceOperationDeclaration";
+        } else {
+          return "operationDeclaration";
+        }
+      case "Interface":
+        return "interfaceDeclaration";
+      case "Enum":
+        return "enumDeclaration";
+      case "EnumMember":
+        return "enumMember";
+      case "Union":
+        if (!type.name) {
+          return "unionLiteral";
+        }
+
+        if (type.templateMapper) {
+          return "unionInstantiation";
+        }
+
+        return "unionDeclaration";
+      case "UnionVariant":
+        return "unionVariant";
+      case "Tuple":
+        return "tupleLiteral";
+      case "Scalar":
+        return "scalarDeclaration";
+      case "Intrinsic":
+        return "intrinsic";
+      default:
+        compilerAssert(false, `Encountered type ${type.kind} which we don't know how to emit.`);
+    }
+  }
+  function currentScope() {
+    return context.referenceContext?.scope ?? context.lexicalContext?.scope ?? null;
+  }
+}
+
+function isDeclaration(type: Type): type is CadlDeclaration | Namespace {
+  switch (type.kind) {
+    case "Namespace":
+    case "Interface":
+    case "Enum":
+    case "Operation":
+    case "Scalar":
+      return true;
+
+    case "Model":
+      return type.name ? type.name !== "" && type.name !== "Array" : false;
+    case "Union":
+      return type.name ? type.name !== "" : false;
+    default:
+      return false;
+  }
+}
+
+/**
+ * An interner takes an object and returns either that same object, or a
+ * previously seen object that has the identical shape.
+ *
+ * This implementation EXTREMELY non-optimal (O(n*m) where n = number of unique
+ * state objects and m = the number of properties a state object contains). This
+ * will very quickly be a bottleneck. That said, the common case is no state at
+ * all, and also this is essentially implementing records and tuples, so could
+ * probably adopt those when they are released. That that said, the records and
+ * tuples are presently facing headwinds due to implementations facing exactly
+ * these performance characteristics. Regardless, there are optimizations we
+ * could consider.
```

Comment:
Worth filing a tracking bug to optimize or do you think it is OK to wait for the bottleneck to arise in practice?

Reply:
I would guess this is in the bucket of things that are slow but unlikely to show up as important behind other slow things except for severely degenerate cases that may not show up in practice.

Reply:
IOW, I would say let's not track. Just something to keep in mind!